### PR TITLE
#96: Added unified error codes

### LINKED ISSUE #94: Add notice about Spark streaming
Acceptance Criteria

- Added remark that Spark streaming not supported yet

### LINKED ISSUE #96: Add unified error handling
## Acceptance Criteria

 - Error messages are unified with code names

### PR DESCRIPTION
**Pull request by morazow**
Fixes #96
Fixes #94

### CODE REVIEW

File: src/main/scala/com/exasol/spark/DefaultSource.scala [hunk 1]
```diff
@@ -112,10 +113,16 @@ class DefaultSource
...
-            s"""|Table $tableName already exists. And DataFrame write mode is set to
-                |`errorifexists` or `default`. Please use one of other SaveMode
-                |modes: 'append', 'overwrite' or 'ignore'.
-            """.stripMargin
+            ExaError
+              .messageBuilder("E-SEC-3")
+              .message(
+                "Table {{TABLE}} already exists in 'errorifexists' or 'default' write modes."
+              )
+              .parameter("TABLE", tableName)
```
Comment (AnastasiiaSergienko) [inline src/main/scala/com/exasol/spark/DefaultSource.scala:121]:
```suggestion
                "Table {{TABLE}} already exists in 'errorifexists' or 'default' write modes.", tableName)
```

Comment (AnastasiiaSergienko) [reply src/main/scala/com/exasol/spark/DefaultSource.scala:121]: You can use a more compact way to add parameters, but the current is also fine if you like it more

File: src/main/scala/com/exasol/spark/rdd/ExasolRDD.scala [hunk 2]
```diff
@@ -56,7 +57,13 @@ class ExasolRDD(
     val conn = manager.mainConnection()
     if (conn == null) {
       logError("Main EXAConnection is null!")
-      throw new RuntimeException("Could not establish main connection to Exasol!")
+      throw new RuntimeException(
+        ExaError
+          .messageBuilder("F-SEC-11")
+          .message("Could not establish main JDBC connection for query.")
+          .mitigation("Please make sure that there network connection between Spark and Exasol clusters.")
```
Comment (AnastasiiaSergienko) [inline src/main/scala/com/exasol/spark/rdd/ExasolRDD.scala:64]:
```suggestion
          .mitigation("Please make sure that there is a network connection between Spark and Exasol clusters.")
```

File: src/main/scala/com/exasol/spark/util/Types.scala [hunk 3]
```diff
@@ -177,10 +178,16 @@ object Types extends Logging {
...
     case TimestampType   => "TIMESTAMP"
-    case _ =>
-      throw new IllegalArgumentException(s"Unsupported Spark data type $dataType!")
+    case _               => throw new IllegalArgumentException(getUnsupportedDataTypeErrorCode(dataType))
   }
 
+  private[this] def getUnsupportedDataTypeErrorCode(dataType: DataType): String =
+    ExaError
+      .messageBuilder("F-SEC-8")
+      .message("Unsupported Spark data type {{DATA_TYPE}} is not supported.", String.valueOf(dataType))
```
Comment (AnastasiiaSergienko) [inline src/main/scala/com/exasol/spark/util/Types.scala:187]: Unsupported data type is not supported -> this sounds funny :)) Maybe rephrase?

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]
Left a few comments. Please also check a/the articles in your error messages. There are too few of them for my taste :)

File: src/main/scala/com/exasol/spark/util/Types.scala (hunk 3 above)
Comment (morazow) [inline src/main/scala/com/exasol/spark/util/Types.scala:187]:
```suggestion
      .message("Spark data type {{DATA_TYPE}} is not supported.", String.valueOf(dataType))
```

Comment (morazow) [reply src/main/scala/com/exasol/spark/util/Types.scala:187]: Typo from previous message, fixed it.

File: src/main/scala/com/exasol/spark/DefaultSource.scala (hunk 1 above)
Comment (morazow) [reply src/main/scala/com/exasol/spark/DefaultSource.scala:121]: Done. It is definitely better for single parameters.

File: doc/changes/changes_1.2.0.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Spark Exasol Connector 1.2.0, released 2021-??-??
+
+Code name:
+
+## Summary
+
+## Refactorings
```
[the remark this answers is not in the document]
Comment (morazow) [reply doc/changes/changes_1.2.0.md:7]: Fixed

**Comment by morazow** [comment]
Links checker fails with lots 429 (too many requests), I am hoping it is fixed automatically later :crossed_fingers:

**Review by AnastasiiaSergienko — APPROVED** [review]
:+1:
